solvation_analysis/analysis.py

```diff
@@ -0,0 +1,127 @@
+        """
+        super(Solute, self).__init__(solute.universe.trajectory, **kwargs)
+        self.radii = {} if radii is None else radii
+        self.kernel = identify_solvation_cutoff if rdf_kernel is None else rdf_kernel
+        # if not kernel:
+        #     self.kernel = identify_solvation_cutoff
+        self.kernel_kwargs = {} if kernel_kwargs is None else radii
+        # if not kernel_kwargs:
+        #     self.kernel_kwargs = {}
+
+        self.solute = solute
+        self.solvents = solvents
+        self.u = self.solute.universe
+        self.rdf_plots = {}
+        self.rdf_data = {}
+        self.ion_speciation = None
+        self.ion_pairing = None
+        self.coordination_numbers = None
+        self.solvation_frames = []
+
+    @classmethod
+    def _plot_solvation_radius(cls, bins, data, radius):
+        fig, ax = plt.subplots()
+        ax.plot(bins, data, "b-", label="rdf")
+        ax.axvline(radius, color="r", label="solvation radius")
+        ax.set_xlabel("Radial Distance (A)")
+        ax.set_ylabel("Probability Density")
+        ax.legend()
+        return fig, ax
+
+    def run_prepare(self):
+        for name, solvent in self.solvents.items():
+            # generate and save RDFs
+            rdf = InterRDF(self.solute, solvent, range=(0.0, 8.0))
+            # TODO: specify start and stop args
+            rdf.run()
+            bins, data = rdf.results.bins, rdf.results.rdf
+            self.rdf_data[name] = (data, bins)
+            # generate and save plots
+            if name not in self.radii.keys():
+                self.radii[name] = self.kernel(bins, data, **self.kernel_kwargs)
+            fig, ax = self._plot_solvation_radius(bins, data, self.radii[name])
+            ax.set_title(f"Solvation distance of {name}")
+            self.rdf_plots[name] = fig, ax
+
+    def _prepare(self):
+        assert self.solvents.keys() == self.radii.keys(), "Radii missing."
+        # columns: solute #, atom id, distance, solvent name, res id
+
+    def _single_frame(self):
+        # initialize empty arrays
+        all_pairs = np.empty((0, 2), dtype=np.int)
+        all_dist = np.empty(0)
+        all_tags = np.empty(0)
+        for name, solvent in self.solvents.items():
+            pairs, dist = distances.capped_distance(
+                self.solute.positions,
+                solvent.positions,
+                self.radii[name],
+                box=self.u.dimensions,
+            )
+            # replace local ids with absolute ids
+            pairs[:, 1] = solvent.ids[[pairs[:, 1]]]  # TODO: ids vs ix?
+            # extend
+            all_pairs = np.concatenate((all_pairs, pairs))
+            all_dist = np.concatenate((all_dist, dist))
+            all_tags = np.concatenate((all_tags, np.full(len(dist), name)))
+        # put the data into a data frame
+        all_resid = self.u.atoms[all_pairs[:, 1]].resids
+        solvation_data_np = np.column_stack(
```

Comment:
instead of making a dataframe I would make  these class data objects like `self.pairs`, `self.distances`,  `self.resnames` and `self.resids` then the user can access the individually without the need to use pandas.

You can then move the pandas stuff to a nice `Solute.to_dataframe()` method that can return the dataframe from the class data.

Reply:
This is definitely the direction I plan to head. I don't intend to expose any pandas directly to the user. My intent is to run the internals with pandas and then have nice accessor functions that can get 90% of the information someone would want.

Reply:
My question is i guess why write accessor functions to cols of a dataframe (basically just a 1D numpy array) when they could just access the array as a class data member directly?

Reply:
I'm thinking the accessors wouldn't be for raw data, only processed. So maybe it doesn't address your concern.

More generally, I don't feel that the raw pairs, resnames, distances, are at all useful. To do pretty much anything the methods would need the solvent_atom_ids, solute_atom_ids, and resnames or resids. Further, they'd all need to share an index or be exactly the same length. I could split them up and then recombine them when needed but that doesn't feel much different than passing through a dataframe. It is more explicit, which perhaps is better.

Reply:
Sharing an index is good enough for the vast majority of cases IMO. AFAIK most Scipy functions that return more than one thing only make them relational via sharing an index.